docs/source/llms/notebooks/chat-model-tool-calling.ipynb

```diff
@@ -0,0 +1,467 @@
+    "        }\n",
+    "    ),\n",
+    "    # make sure to call `to_tool_definition()` to convert the `FunctionToolDefinition`\n",
+    "    # to a `ToolDefinition` object. this step is necessary to normalize the data format,\n",
+    "    # as multiple types of tools (besides just functions) might be available in the future.\n",
+    ").to_tool_definition()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 2: Implementing the tool\n",
+    "\n",
+    "Now that we have a definition for the tool, we need to actually implement it. For the purposes of this tutorial, we're just going to mock a response, but the implementation can be arbitrary—you might make an API call to an actual weather service, for example."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_weather(city: str) -> str:\n",
+    "    return f\"It's sunny in {city}, with a temperature of 20C\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Step 3: Defining our model\n",
+    "\n",
+    "Now we're ready to create our model! We'll be subclassing `mlflow.pyfunc.ChatModel`, and this involves defining a `predict()` function that accepts the following arguments:\n",
+    "\n",
+    "1. `context`: [PythonModelContext](https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.PythonModelContext) (not used in this tutorial)\n",
+    "2. `messages`: List\\[[ChatMessage](https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.llm.ChatMessage)\\]. This is the chat input that the model uses for generation.\n",
+    "3. `params`: [ChatParams](https://mlflow.org/docs/latest/python_api/mlflow.types.html#mlflow.types.llm.ChatParams). These are commonly used params used to configure the chat model, e.g. `temperature`, `max_tokens`, etc. This is where the tool specifications can be found.\n",
+    "\n",
+    "For the implementation, we'll simply forward the user's input to OpenAI, and provide the `get_weather` tool as an option for the LLM to use if it chooses to do so. If we receive a tool call request, we'll call the `get_weather()` function and return the response back to OpenAI. We'll need to use what we've defined in the previous two steps in order to do this."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "from typing import List\n",
+    "\n",
+    "from openai import OpenAI\n",
+    "\n",
+    "import mlflow\n",
+    "from mlflow.types.llm import (\n",
+    "    ChatMessage,\n",
+    "    ChatParams,\n",
+    "    ChatResponse,\n",
+    "    FunctionToolDefinition,\n",
+    "    ParamProperty,\n",
+    "    ToolParamsSchema,\n",
+    ")\n",
+    "\n",
+    "# replace this with your own MLflow tracking URI\n",
+    "mlflow.set_tracking_uri(\"http://localhost:5000\")\n",
```

Comment:
Can we add a step to launch a tracking server?

Reply:
done!